### ISSUE TITLE
JerseyClientConfiguration.setGzipEnabled(false) not working as expected

### ISSUE THREAD

**Issue by lehcim** [dropwizard/dropwizard#1267]
I'm using dropwizard 0.9.0-rc4 to build a service which uses another Rest API. In order to debug my application, I'm trying to observe the requests and responses between a local jersey client and the remote server on a logging proxy. To make the logs human readables, I tried to disable the compression on the client side like this:

```java
JerseyClientConfiguration clientConfig = new JerseyClientConfiguration();
clientConfig.setGzipEnabled(false);
clientConfig.setGzipEnabledForRequests(false);

Client client = new JerseyClientBuilder(environment).using(clientConfig).build("myClient")
```

but the  remote server answered with a compressed answer, because the request was looking like this:

```
GET /foo/bar HTTP/1.1
Host: localhost:8080
Connection: Keep-Alive
User-Agent: myClient
Accept: application/json,application/xml
Accept-Encoding: gzip,deflate
```

Where do this 'Accept-Encoding' header come from, since gzip is disabled in the JerseyClientConfiguration ?
In short, if you disable compression at the Jersey level, then apache will enable it for you again...

* The io.dropwizard.client.JerseyClientBuilder delegates client creation to io.dropwizard.client.HttpClientBuilder, which uses org.apache.http.impl.client.HttpClientBuilder to really create the instance.

* In the apache builder, compression is enabled by default, if you don't call explicitely disableContentCompression() then the "gzip,deflate" is automatically added by org.apache.http.client.protocol.RequestAcceptEncoding ...

* Thus, to honor the .setGzipEnabled(false) call, the dropwizard JerseyClientBuilder should tell the dropwizard HttpClientBuidlder to disable the compression in the apache HttpClientBuilder... but the gzipEnabled flag is not visible in dropwizard HttpClientBuilder, because it is a member of the above JerseyClientConfiguration 

A possible way to solve this issue is to force disableContentCompression() in apache HttpClientBuilder, because Gzip is managed at the Jersey level (see io.dropwizard.jersey.gzip.GzipDecoder). It should work
in my use case, but I can't say it will not break something elsewhere.

Any thoughts ?

**Comment by arteam** [comment]
It makes sense to me. 

I don't think we will break anything, because compression is enabled by default. We probably broke this feature, when we moved to `org.apache.http.impl.client.HttpClientBuilder` approach for constructing the client. It  (de)compresses by default, but `DefaultHttpClient` don't.

**Comment by lehcim** [comment]
I tried to disable the compression in io.dropwizard.client.HttpClientBuilder, at the end of createClient :

```java
protected ConfiguredCloseableHttpClient createClient(
            final org.apache.http.impl.client.HttpClientBuilder builder,
            final InstrumentedHttpClientConnectionManager manager,
            final String name) {
        // ....
        if (routePlanner != null) {
            builder.setRoutePlanner(routePlanner);
        }

        // because Gzip is managed at the Jersey level
        builder.disableContentCompression() ;
        
        return new ConfiguredCloseableHttpClient(builder.build(), requestConfig);
    }
```
and it work as expected with`clientConfig.setGzipEnabled(false);`

However, with `clientConfig.setGzipEnabled(false);`, I don't see the expected `Accept-Encoding: gzip` in the request.... I think the header is to be added by GzipDecoder, see the first `if` statement:

```java
public class GZipDecoder implements ReaderInterceptor {
    @Override
    public Object aroundReadFrom(ReaderInterceptorContext context) throws IOException {
        if (!context.getHeaders().containsKey(HttpHeaders.ACCEPT_ENCODING)) {
            context.getHeaders().add(HttpHeaders.ACCEPT_ENCODING, "gzip");
        }

        String contentEncoding = context.getHeaders().getFirst(HttpHeaders.CONTENT_ENCODING);
        if (contentEncoding != null &&
                (contentEncoding.equals("gzip") || contentEncoding.equals("x-gzip"))) {
            context.setInputStream(new GZIPInputStream(context.getInputStream()));
        }
        return context.proceed();
    }
}
````
but it can't work here, we are in a ReaderInterceptor, we already got the response ... if we really want to add the Accept-Encoding value at the Jersey level, GzipDecoder has to implements ClientRequestFilter as well.

Will test it, and submit a PR if OK.

**Comment by lehcim** [comment]
Here is the pull request: https://github.com/dropwizard/dropwizard/pull/1268

**Comment by arteam** [comment]
Why do you expect `Accept-Encoding: gzip` if we don't want to get compressd responses? Isn't the whole idea to omit this header?

**Comment by arteam** [comment]
See #1270. Does it work as you expect?

**Comment by lehcim** [comment]
Not tested #1270 yet, but after reading the diff I think it should work OK.

About the `Accept-Encoding: gzip` in GZipDecoder: you need to set/unset this header at the Jersey level if you want/don't want compressed responses with an client whose compression was disabled (in my PR, I have not added a boolean in the configuration). 

Anyway, in current GZipDecoder I think the `add(HttpHeaders.ACCEPT_ENCODING, "gzip")`  is dead code : we are in a ReaderInterceptor, not in a ClientRequestFilter, we already have the response, it's too late to say we accept some encoding or not.

**Comment by lehcim** [comment]
Just tested it, and yes, #1270 work as expected.
Thank you !

**Comment by arteam** [comment]
That's great! 

I am waiting for someone from the core team to review the pull request and then it can be merged.

**Comment by arteam** [comment]
Closed via #1270
